kms/src/test/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryKmsServiceTest.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.provider.kroxylicious.inmemory;
+
+import java.nio.ByteBuffer;
+import java.util.ServiceLoader;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.kroxylicious.kms.service.KmsService;
+import io.kroxylicious.kms.service.Ser;
+import io.kroxylicious.kms.service.UnknownAliasException;
+import io.kroxylicious.kms.service.UnknownKeyException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class InMemoryKmsServiceTest {
+
+    InMemoryKmsService service;
+
+    @BeforeEach
+    public void before() {
+        service = createServiceInstance();
+    }
+
+    private static InMemoryKmsService createServiceInstance() {
+        return (InMemoryKmsService) ServiceLoader.load(KmsService.class).stream()
+                .filter(p -> p.type() == InMemoryKmsService.class)
+                .findFirst()
+                .get()
+                .get();
+    }
+
+    @Test
+    void shouldRejectOutOfBoundIvBytes() {
+        // given
+        assertThrows(IllegalArgumentException.class, () -> new InMemoryKmsService.Config(0, 128));
+    }
+
+    @Test
+    void shouldRejectOutOfBoundAuthBits() {
+        assertThrows(IllegalArgumentException.class, () -> new InMemoryKmsService.Config(12, 0));
+    }
+
+    @Test
+    void shouldSerializeAndDeserialiseKeks() {
+        // given
+        var kms = service.buildKms(new InMemoryKmsService.Config(12, 128));
+        var kek = kms.generateKey();
+        assertNotNull(kek);
+
+        // when
+        Ser<UUID> kekSer = kms.keyIdSerializer();
+        var buffer = ByteBuffer.allocate(kekSer.sizeOf(kek));
+        kekSer.serialize(kek, buffer);
+        assertFalse(buffer.hasRemaining());
+        buffer.flip();
+        var kekDe = kms.keyIdDeserializer();
+        var loadedKek = kekDe.deserialize(buffer);
+
+        // then
+        assertEquals(kek, loadedKek, "Expect the deserialized kek to be equal to the original kek");
+    }
+
+    @Test
+    void shouldGenerateDeks() {
+        // given
+        var kms1 = service.buildKms(new InMemoryKmsService.Config(12, 128));
+        var kms2 = createServiceInstance().buildKms(new InMemoryKmsService.Config(12, 128));
+        var key1 = kms1.generateKey();
+        assertNotNull(key1);
+        var key2 = kms2.generateKey();
+        assertNotNull(key2);
+
+        // when
+        CompletableFuture<InMemoryEdek> gen1 = kms1.generateDek(key2);
```

Comment:
I'd use CompletionStage#toCompletableFuture() in the tests and have the impl only return CompletionStages.

Reply:
What's the benefit of doing that? the In-memory KMS will only be used for tests anyway? All the production code will be written in terms of `CompletionStage`, since that's the type used in the KMS API.